squint.c

```diff
@@ -0,0 +1,1609 @@
+/**********************************************************/
+
+/* The ARM processor allows use of pc-relative constants */
+
+/* constants are sorted by the address they are stored in memory */
+int find_const(int addr)
+{
+   int low, high, mid;
+
+   // pool is sorted by *relative address* of data values
+   // with respect to beginning of text segment.
+   // binsearch pool for address of data
+   low = 0; high = cnst_pool_size;
+   while (low != high) {
+      mid = (low + high) / 2;
+      if (cnst_pool[mid].data_addr <= addr) {
+         low = mid;
+         if (cnst_pool[mid].data_addr == addr) break;
+         if (high-low == 1) ++low;
+      }
+      else
+         high = mid;
+   }
+   return low;
+}
+
+int is_const(int *inst)
+{
+   int addr = (inst-cbegin)*4;
+
+   return (cnst_pool[find_const(addr)].data_addr == addr);
+}
+
+/* Amacc pc-relative constants always appear in the instruction */
+/* stream after the instructions that reference them.  Because */
+/* of this we can safely scan an instruction stream forward, */
+/* skipping any constants we encounter. */
+void create_const_map(int *begin, int *end)
+{
+   int *scan = begin;
+   int i;
+
+   /* initialize global variables */
+   cbegin = begin;
+   cnst_pool = calloc(end-begin, sizeof(struct pd_s)); // 1/4 prog size
+   cnst_pool_size = 0;
+
+   while (scan < end) {
+      if (!is_const(scan) &&
+          (*scan & 0xffff0000) == 0xe59f0000) { // ldr r0, [pc, #X]
+         int addr = ((scan + 2 + (*scan & 0xfff) / 4) - begin)*4;
+         struct ia_s **inst;
+         int low = find_const(addr);
+
+         if (cnst_pool[low].data_addr != addr) {
+            int *mem = ((int *)&cnst_pool[cnst_pool_size]) + 1;
+
+            // insert addr at this location
+            for (i = cnst_pool_size; i > low; --i) {
+               *mem = *(mem-2); --mem;
+               *mem = *(mem-2); --mem;
+            }
+            cnst_pool[low].data_addr = addr;
+            cnst_pool[low].inst = 0;
+            ++cnst_pool_size;
+         }
+         /* keep track of pc-relative instructions that ref this constant */
+         for (inst = &cnst_pool[low].inst; *inst != 0; inst = &(*inst)->next);
+         *inst = malloc(sizeof(struct ia_s));
+         (*inst)->inst_addr = (scan-begin)*4;
+         (*inst)->next = 0;
+      }
+      ++scan;
+   }
+}
+
+void destroy_const_map()
+{
+   struct ia_s *inst, *next;
+   int i;
+
+   for (i = 0;  i < cnst_pool_size; ++i) {
+      inst = cnst_pool[i].inst;
+      while (inst != 0) {
+         next = inst->next;
+         free(inst);
+         inst = next;
+      }
+   }
+   free(cnst_pool);
+}
+
+/* convert special ARM consts to mov immediate */
+void const_imm_opt(int *begin, int *end)
+{
+   struct ia_s *inst, *next;
+   int *newinst;
+   int i, j, val, rotate;
+
+   for (i = 0, j = 0; i < cnst_pool_size; ++i) {
+      val = cbegin[cnst_pool[i].data_addr/4];
+      rotate = 0;
+      if ((val & 0x0000ff00) && !(val & 0xffff00ff)) rotate = 1;
```

Comment:
This can be generalized by a function that can match all possible modified immediate:

```
// Returns the ARM modified immediate if x can be encoded as a modified
// immediate; otherwise, return -1.
int encode_arm_modified_immediate(unsigned x) {
  int rotate;
  for (rotate = 0; rotate < 16; ++rotate) {
    if ((x & 0xffffff00) == 0) {
      return (rotate << 8) | x;
    }
    x = (x << 2) | (x >> 30);  // Rotate 2 bits from MSB to LSB
  }
  return -1;
}
```

(p.s. It is possible to further improve the performance with `__builtin_clz` and `__builtin_ctz`, but I don't have time to think about it thoroughly.)

And then check the constant value with:

```
int mov_imm_inst = 0;
int immediate;
if ((immediate = encode_arm_modified_constant(val)) != -1) {
  mov_imm_inst = 0xe3a00000;  // mov
} else if ((immediate = encoding_arm_modified_constant(~val)) != -1) {
  mov_imm_inst = 0xe3e00000;  // mvn
}
if (mov_imm_inst) {
  inst = cnst_pool[i].inst;
  while (inst != 0) {
    next = inst->next;
    newinst = cbegin + inst->inst_addr/4;
    *newinst = mov_imm_inst | (*newinst & RI_Rd) | immediate;
    free(inst);
  }
}
```

Reply:
x = (x << 2) | (x >> 30);  // Rotate 2 bits from MSB to LSB

I believe 0x80000000 >> 30 will smear a 1 through the top 30 bits, corrupting x.

PS  I don't mind the suggested peephole checks being changed after this pull request, or any other changes. I just won't be the one to do it.

Reply:
> I believe 0x80000000 >> 30 will smear a 1 through the top 30 bits, corrupting x.

I used `unsigned` as the type for `x`.  `x >> 30` should be translated to logical shift right (LSR), which should pad zeros instead of the most significant bit (MSB).

Reply:
You have the option to discard this pull request with no ill feelings on my part.  I suspected this might be too big for AMaCC when I started, which I was hesitant in my first post for this pull request.

In any event, I will continue to submit small bug fix pull requests to AMaCC that touch 20 or less lines.  I will return to fixing my amacc "issue" submissions  around March 14, after I have completed float data type support for this code base on my personal repo @ https://github.com/HPCguy/Squint.

Reply:
When I fix issue #94 in AMaCC, amacc will be able to compile Squint, at least in my repo.  I'm hoping that if I add optimizations to Squint that beat gcc -O1 on a wide variety of tests, you guys might accept some Squint changes back, assuming you guys were to accept this pull request.  I think I'm hoping to beat gcc -O1 on a variety of tests with as little as 500 lines of additional optimizations in Squint.